In [2]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [3]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [4]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

## 1. Load Data & JOIN

*Joining all four tables — `return_data`, `master_sku`, `invoice_data`, and `master_customer` — into a single analytical table using SQL JOIN. This combined table will be used throughout the entire cleaning and analysis process.*

In [13]:
%%sql

SELECT
    rd.retur_id,
    id.invoice_id,
    ms.sku_id,
    mc.customer_id,
    rd.return_date,
    rd.return_reason,
    ms.sku_name,
    ms.kategori,
    mc.customer_type,
    mc.region,
    rd.value_return
FROM
    return_data rd
LEFT JOIN
    master_sku ms on ms.sku_id = rd.sku_id 
LEFT JOIN
    invoice_data id on id.invoice_id = rd.invoice_id
LEFT JOIN
    master_customer mc on mc.customer_id = id.customer_id
LIMIT 5

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,return_reason,sku_name,kategori,customer_type,region,value_return
RT00028,009926,R011,RSS066,2024-09-04,Salah Pesan Customer,Selimut Pasien Rumah Sakit,Perawatan & Rawat Inap,RS Swasta,Bekasi,657000.0
RT00043,010033,D013,RSS037,2024-07-17,Kadaluarsa,Stethoscope Anak,Diagnostik,RS Swasta,Jakarta Barat,2761000.0
RT00002,009510,B014,RSP030,2024-07-21,Salah Pesan Customer,Duk Operasi 80x90cm,Bedah & Tindakan,RS Pemerintah,Bekasi,1067000.0
RT00053,010143,D007,RSS004,2024-06-05,Salah Input Sistem,Rapid Test HbsAg,Diagnostik,RS Swasta,Jakarta Barat,8195000.0
RT00058,010180,D015,RSS019,2024-11-16,Salah Pesan Customer,Penlight Diagnostik,Diagnostik,RS Swasta,Jakarta Barat,364000.0


In [14]:
%%sql

SELECT
    COUNT(*) AS total_records
FROM
    return_data rd
LEFT JOIN
    master_sku ms on ms.sku_id = rd.sku_id 
LEFT JOIN
    invoice_data id on id.invoice_id = rd.invoice_id
LEFT JOIN
    master_customer mc on mc.customer_id = id.customer_id

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
61


In [15]:
%%sql

SELECT
    COUNT(*) AS total_records
FROM
    return_data rd

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
59


**Finding:**  
Hasil JOIN keempat tabel menghasilkan 61 baris, lebih banyak dari jumlah asli `return_data` (59 baris). Selisih 2 baris ini disebabkan oleh duplikat murni pada `invoice_data` yang ditemukan di tahap Data Understanding — setiap `invoice_id` yang terduplikasi menyebabkan baris `return_data` yang terhubung ikut terduplikasi saat JOIN.

In [16]:
%%sql

WITH unique_invoices AS (
    SELECT
        DISTINCT *
    FROM
        invoice_data
)

SELECT
    rd.retur_id,
    ui.invoice_id,
    ms.sku_id,
    mc.customer_id,
    rd.return_date,
    rd.return_reason,
    ms.sku_name,
    ms.kategori,
    mc.customer_type,
    mc.region,
    rd.value_return
FROM
    return_data rd
LEFT JOIN
    master_sku ms on ms.sku_id = rd.sku_id 
LEFT JOIN
    unique_invoices ui on ui.invoice_id = rd.invoice_id
LEFT JOIN
    master_customer mc on mc.customer_id = ui.customer_id
LIMIT 5

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,return_reason,sku_name,kategori,customer_type,region,value_return
RT00042,010029,D001,RSP053,2024-03-02,Barang Rusak,Tensimeter Digital Lengan,Diagnostik,RS Pemerintah,Jakarta Utara,1468000.0
RT00046,010092,B002,RSS044,2024-05-14,Salah Pesan Customer,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,RS Swasta,Jkt Selatan,417500.0
RT00021,009825,L002,RSS056,2024-11-22,Salah Input Sistem,Tabung Vacutainer Plain,Laboratorium,RS Swasta,Bekasi,478000.0
RT00023,009861,R007,KLN012,2024-12-23,Salah Pesan Customer,Tiang Infus Stainless 5-Roda,Perawatan & Rawat Inap,Klinik,Bogor,6207500.0
RT00027,009902,C020,KLN059,2024-08-06,Salah Input Sistem,Spalk Lengan Anak,Consumable,Klinik,Jakarta Timur,252000.0


In [18]:
%%sql

WITH unique_invoices AS (
    SELECT
        DISTINCT *
    FROM
        invoice_data
)

SELECT
    COUNT(*) AS total_records
FROM
    return_data rd
LEFT JOIN
    master_sku ms on ms.sku_id = rd.sku_id 
LEFT JOIN
    unique_invoices ui on ui.invoice_id = rd.invoice_id
LEFT JOIN
    master_customer mc on mc.customer_id = ui.customer_id
LIMIT 5

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
59


**Finding:**  
Hasil JOIN awal menghasilkan 61 baris akibat duplikat pada `invoice_data`. Setelah menerapkan `SELECT DISTINCT` melalui `CTE` `invoice_clean`, jumlah baris kembali sesuai dengan jumlah asli `return_data`, yaitu 59 baris.

**Conclusion:**  
Analytical table berhasil dibuat dengan benar — setiap retur tercatat tepat satu kali, tanpa duplikasi akibat data quality issue pada `invoice_data`.

### Create Analytical View

*Creating a SQL `VIEW` that joins all four tables — `return_data`, `master_sku`, `invoice_data`, and `master_customer` — into a single analytical table. A CTE (`invoice_clean`) is used inside the view to remove duplicate records found in `invoice_data` during Data Understanding. This view will be reused throughout the rest of the cleaning and analysis process, avoiding the need to rewrite the JOIN query in every subsequent section.*

In [9]:
%%sql

CREATE VIEW return_analysis AS
WITH unique_invoices AS (
    SELECT
        DISTINCT *
    FROM
        invoice_data
)

SELECT
    rd.retur_id,
    ui.invoice_id,
    ms.sku_id,
    mc.customer_id,
    rd.return_date,
    rd.return_reason,
    ms.sku_name,
    ms.kategori,
    mc.customer_type,
    mc.region,
    rd.value_return
FROM
    return_data rd
LEFT JOIN
    master_sku ms on ms.sku_id = rd.sku_id 
LEFT JOIN
    unique_invoices ui on ui.invoice_id = rd.invoice_id
LEFT JOIN
    master_customer mc on mc.customer_id = ui.customer_id


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
Done.


[]

In [10]:
%%sql
SELECT 
    COUNT(*) AS total_rows 
FROM 
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_rows
59


**Finding:**  
View `return_analysis` berhasil dibuat dengan menggabungkan keempat tabel. Total baris terkonfirmasi sebanyak 59, sesuai dengan jumlah asli `return_data` — mengonfirmasi bahwa duplikat pada `invoice_data` sudah berhasil ditangani melalui CTE `unique_invoices`.

**Conclusion:**  
View ini akan menjadi sumber data utama untuk seluruh tahap Data Cleaning dan EDA selanjutnya.

## 2. Handle Duplicates

*Verifying that the analytical view (`return_analysis`) is free from duplicate records, confirming the deduplication applied during the JOIN process in Section 1 was successful.*

In [12]:
%%sql

SELECT
    retur_id,
    COUNT(*) AS total_returns
FROM
    return_analysis
GROUP BY
    retur_id
HAVING
    COUNT(*) > 1

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


retur_id,total_returns


**Finding:**  
Tidak ditemukan `retur_id` yang terduplikasi pada `return_analysis`. Deduplikasi `invoice_data` melalui `CTE` pada Section 1 berhasil mencegah duplikasi data saat proses JOIN.

**Conclusion:**  
View `return_analysis` sudah bebas dari duplikat dan siap digunakan untuk tahap pembersihan data selanjutnya.

## 3. Handle Missing Value

*Checking for missing values in `return_analysis` — particularly on columns derived from the JOIN (`kategori`, `customer_type`, `region`) — to confirm that every foreign key successfully matched its reference table.*

In [13]:
%%sql

SELECT
    COUNT(*) - COUNT(kategori) AS missing_kategori,
    COUNT(*) - COUNT(customer_type) AS missing_customer_type,
    COUNT(*) - COUNT(region) AS missing_region
FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_kategori,missing_customer_type,missing_region
0,0,0


**Finding:**  
Tidak ditemukan missing value pada kolom (`kategori`, `customer_type`, `region`) pada `return_analysis` setelah di lakukan JOIN.

**Conclusion:**  
View `return_analysis` bebas dari missing values, mengonfirmasi seluruh foreign key (`sku_id`, `invoice_id`, `customer_id`) berhasil matched dengan tabel referensinya masing-masing. Data siap untuk tahap standarisasi format selanjutnya.

## 4. Standardize Formatting

*Standardizing inconsistent text formatting found during Data Understanding — including mixed casing and abbreviations on `return_reason`, `customer_type`, and `region`. A new view (`return_clean`) is created on top of `return_analysis`, applying `CASE WHEN` logic to unify each inconsistent value into a single standard category.*

In [42]:
%%sql

SELECT
    DISTINCT return_reason
FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
6 rows affected.


return_reason
Slh Input
Salah Input Sistem
customer salah pesan
Barang Rusak
Kadaluarsa
Salah Pesan Customer


In [17]:
%%sql

SELECT
    DISTINCT customer_type
FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


customer_type
Klinik
RS Pemerintah
klinik
RS SWASTA
RS Swasta


In [15]:
%%sql

SELECT
    DISTINCT region
FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
11 rows affected.


region
JKT PUSAT
Jkt Selatan
Jakarta Pusat
Tangerang
Jakarta Utara
Bekasi
Depok
Jakarta Barat
Bogor
Jakarta Selatan


In [20]:
%%sql

SELECT
    CASE
        WHEN return_reason = 'customer salah pesan' THEN 'Salah Pesan Customer'
        WHEN return_reason = 'Slh Input' THEN 'Salah Input Sistem'
        ELSE return_reason
    END AS return_reason_cleaned,

    CASE
        WHEN customer_type = 'RS SWASTA' THEN 'RS Swasta'
        WHEN customer_type = 'klinik' THEN 'Klinik'
        ELSE customer_type
    END AS customer_type_cleaned,

    CASE
        WHEN region = 'Jkt Selatan' THEN 'Jakarta Selatan'
        WHEN region = 'JKT PUSAT' THEN 'Jakarta Pusat'

        ELSE region
    END AS region_cleaned

FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
59 rows affected.


return_reason_cleaned,customer_type_cleaned,region_cleaned
Barang Rusak,RS Pemerintah,Jakarta Utara
Salah Pesan Customer,RS Swasta,Jakarta Selatan
Salah Input Sistem,RS Swasta,Bekasi
Salah Pesan Customer,Klinik,Bogor
Salah Input Sistem,Klinik,Jakarta Timur
Salah Pesan Customer,Klinik,Jakarta Barat
Salah Input Sistem,RS Pemerintah,Bekasi
Salah Input Sistem,RS Swasta,Jakarta Barat
Salah Input Sistem,RS Swasta,Jakarta Barat
Salah Pesan Customer,Klinik,Bekasi


In [ ]:
%%sql

CREATE VIEW return_clean AS

SELECT
    retur_id,
    invoice_id,
    sku_id,
    customer_id,
    return_date,
    sku_name,
    kategori,
    value_return,

    CASE
        WHEN return_reason = 'customer salah pesan' THEN 'Salah Pesan Customer'
        WHEN return_reason = 'Slh Input' THEN 'Salah Input Sistem'
        ELSE return_reason
    END AS return_reason_cleaned,

    CASE
        WHEN customer_type = 'RS SWASTA' THEN 'RS Swasta'
        WHEN customer_type = 'klinik' THEN 'Klinik'
        ELSE customer_type
    END AS customer_type_cleaned,

    CASE
        WHEN region = 'Jkt Selatan' THEN 'Jakarta Selatan'
        WHEN region = 'JKT PUSAT' THEN 'Jakarta Pusat'
        ELSE region
    END AS region_cleaned    
    
FROM
    return_analysis

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
Done.


[]

In [39]:
%%sql

SELECT
    DISTINCT return_reason_cleaned
FROM
    return_clean

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned
Salah Input Sistem
Barang Rusak
Kadaluarsa
Salah Pesan Customer


In [40]:
%%sql

SELECT
    DISTINCT customer_type_cleaned
FROM
    return_clean

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned
Klinik
RS Pemerintah
RS Swasta


In [41]:
%%sql

SELECT
    DISTINCT region_cleaned
FROM
    return_clean

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned
Jakarta Pusat
Tangerang
Jakarta Utara
Bekasi
Depok
Jakarta Barat
Bogor
Jakarta Selatan
Jakarta Timur


**Finding:**  
setelah di lakukan join, di temukan:
- 6 kategori di return reason dimana 2 di antaranya tidak standar
- 5 kategori di customer_type yang 2 di antaranya tidak standar
- 11 kategori di region yang 2 di antaranya tidak standar  

setelah di lakukan standarisasi dengan Case when, kategori yang tidak standar, di ubah menjadi 1 standar yang sma, hasilnya:
- 4 kategori pada return sesason sudah standar
- 3 kategori di customer_type sudah standar
- 9 kategori pada region sudah standar

**Conclusion:**  
View `return_clean` bebas dari data yang tidak standar dan siap digunakan untuk tahap analisis outlier dan EDA selanjutnya.

## 5. Handle Outliers

*Checking for outliers in `value_return` using the IQR method — calculating Q1, Q3, and the lower/upper bounds with `PERCENTILE_CONT()` to identify return transactions with extreme loss values.*

In [54]:
%%sql

WITH iqr_count AS
(
    SELECT
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY value_return) AS Q1,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY value_return) AS Q3
FROM
    return_clean
)


SELECT
    Q1,
    Q3,
    (Q3-Q1) AS IQR_value,
    Q1 - ((Q3-Q1) * 1.5) AS lower_bound,
    Q3 + ((Q3-Q1) * 1.5) As Upper_bound
FROM
    iqr_count

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


q1,q3,iqr_value,lower_bound,upper_bound
422250.0,1896500.0,1474250.0,-1789125.0,4107875.0


In [62]:
%%sql


CREATE VIEW return_cleaned AS
(
    WITH iqr_count AS
(
    SELECT
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY value_return) AS Q1,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY value_return) AS Q3
FROM
    return_clean
)
    
    SELECT
        rc.*,
        (ic.Q3 + ((Q3-Q1) * 1.5)) AS upper_bound
    FROM
        return_clean rc
    CROSS JOIN
        iqr_count AS ic
)


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
Done.


[]

In [65]:
%%sql

SELECT
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    value_return > upper_bound

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
12


In [66]:
%%sql

SELECT * 
FROM
    return_cleaned
LIMIT 5

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0
RT00023,009861,R007,KLN012,2024-12-23,Tiang Infus Stainless 5-Roda,Perawatan & Rawat Inap,6207500.0,Salah Pesan Customer,Klinik,Bogor,4107875.0
RT00027,009902,C020,KLN059,2024-08-06,Spalk Lengan Anak,Consumable,252000.0,Salah Input Sistem,Klinik,Jakarta Timur,4107875.0


In [69]:
%%sql

SELECT
    retur_id,
    return_date,
    sku_name,
    kategori,
    value_return,
    return_reason_cleaned,
    customer_type_cleaned,
    region_cleaned
FROM
    return_cleaned
WHERE
    value_return > upper_bound

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
12 rows affected.


retur_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned
RT00023,2024-12-23,Tiang Infus Stainless 5-Roda,Perawatan & Rawat Inap,6207500.0,Salah Pesan Customer,Klinik,Bogor
RT00053,2024-06-05,Rapid Test HbsAg,Diagnostik,8195000.0,Salah Input Sistem,RS Swasta,Jakarta Barat
RT00055,2024-07-04,Pulse Oximeter,Diagnostik,4650000.0,Salah Input Sistem,RS Swasta,Jakarta Barat
RT00037,2024-06-16,Urine Stick 10 Parameter,Diagnostik,18473000.0,Salah Pesan Customer,Klinik,Bekasi
RT00036,2024-04-25,Benang Catgut 2-0,Bedah & Tindakan,5537000.0,Salah Pesan Customer,RS Swasta,Jakarta Utara
RT00020,2024-03-10,IV Catheter 22G,Consumable,14563500.0,Salah Pesan Customer,RS Pemerintah,Bogor
RT00012,2024-08-24,Rapid Test HbsAg,Diagnostik,15979000.0,Barang Rusak,RS Pemerintah,Tangerang
RT00019,2024-05-06,Glukometer Set,Diagnostik,6300000.0,Salah Input Sistem,Klinik,Jakarta Selatan
RT00030,2024-10-02,Handscoon Steril 7.5,Bedah & Tindakan,12801000.0,Salah Pesan Customer,RS Swasta,Bekasi
RT00035,2024-05-09,Benang Catgut 2-0,Bedah & Tindakan,10678500.0,Salah Input Sistem,RS Swasta,Jakarta Utara


**Finding:**  
Berdasarkan perhitungan IQR diketahui
- `Q1` = 422250.0
- `Q3` = 1896500
- `IQR` = 1474250
- `lower_bound` = -1789125
- `upper_bound` = 4107875  

Serta di temukan 12 data outlier atau di atas upper bound

**Conclusion:**  
- Berdasarkan data yang diketahui, `lower_bound` dengan nilai negatif tidak mungkin dijadikan acuan analisis, karena `value_return` secara logis tidak bisa bernilai negatif. 
- Berdasarkan `upper_bound`, ditemukan 12 data yang melebihi batas tersebut. 
- Ke-12 data ini akan dipertahankan untuk analisis selanjutnya di EDA, karena merupakan transaksi retur yang valid (bukan data error) dan relevan untuk menjawab Q3 — total nilai kerugian akibat retur.

## 6. Export to CSV

In [70]:
df_return = pd.read_sql("SELECT * FROM return_cleaned", connection)

In [72]:
df_return.to_csv('return_clean.csv', index=False)

In [73]:
df_return.shape

(59, 12)

In [75]:
df_return.head(5)

,retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
0,RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
1,RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
2,RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0
3,RT00023,009861,R007,KLN012,2024-12-23,Tiang Infus Stainless 5-Roda,Perawatan & Rawat Inap,6207500.0,Salah Pesan Customer,Klinik,Bogor,4107875.0
4,RT00027,009902,C020,KLN059,2024-08-06,Spalk Lengan Anak,Consumable,252000.0,Salah Input Sistem,Klinik,Jakarta Timur,4107875.0


**Finding:**  
Data hasil cleaning berhasil di-export menjadi `return_clean.csv` yang terdiri dari 59 baris dan 12 kolom.

**Conclusion:**  
Dataset ini siap digunakan untuk tahap EDA selanjutnya, mencakup seluruh kolom yang relevan untuk menjawab *Business Questions*.